### ISSUE TITLE
Tgav constraint

### ISSUE THREAD

**Issue by bpbond** [JGCRI/hector#557]
@kdorheim asks
>Do we know if the Hector tgav constraint actually works?

I am opening this issue mostly for the pleasure of closing it.

**Comment by kdorheim** [comment]
1 min

**Comment by kdorheim** [comment]
Okay as part of the #558 could you go through the section of the [temperature component](https://github.com/JGCRI/hector/blob/2d2c9d36197b30f460e34d74d3b3e5e60736a099/src/temperature_component.cpp#L352) and clean up inline code documentation? There are some comments like 

>  // Temperature component can't handle that at the moment!
        H_LOG( logger, Logger::SEVERE ) << "** ERROR - Temperature can't currently handle user-supplied temperature" << std::endl;
        H_THROW("User-supplied temperature not yet implemented.")

which might be useful tests but also are confusing and what caused me to question if the tgav constraint actually worked or not.

**Comment by bpbond** [comment]
@kdorheim Here's what used to happen, pre-DOECLIM when the temperature constraint code got nuked:
```cpp
// We track total radiative forcing using internal variable `internal_Ftot`
// Need to do this because if we're subject to a user constraint (being forced
// to match a temperature record), need to track the Ftot that *would* have
// produced the observed temperature record. This way there's a smooth
// transition when we exit the constraint period, after which internal_Ftot
// will rise in parallel with the value reported by ForcingComponent.

// If we never had any temperature constraint, `internal_Ftot` will match `Ftot`.
// If the user has supplied temperature data, use that (except if past its end)
if( tgav_constrain.size() && runToDate <= tgav_constrain.lastdate() ) {
	H_LOG( logger, Logger::NOTICE ) << "** Using user-supplied temperature" << std::endl;
	tgav = tgav_constrain.get( runToDate );
	tgaveq = tgav;

	// Update last_Ftot and internal_Ftot for when we exit constraint
	last_Ftot = core->sendMessage( M_GETDATA, D_RF_TOTAL ).value( U_W_M2 );
	const double heatflux = core->sendMessage( M_GETDATA, D_HEAT_FLUX ).value( U_W_M2 );
	internal_Ftot = 3.7 * tgav / S + heatflux;  // inverting normal T=S/3.7*F eqn
} else {  // not constrained
```
...and then there's a lot complicated crap whereby we used `internal_Ftot` to ensure that when we exited a temperature constraint we smoothly transitioned back to the unconstrained behavior.

I'm inclined not to worry about this last part: if the model jumps when exiting a constraint, that's the user's problem, not ours. Thoughts?

For the record, this is based on bd1ae8901c2a69a10c75feefeee71e31be83d7e0 from 2016.

**Comment by kdorheim** [comment]
Re the model jumps - what do we do with the other constraints? Do we smooth it out or let the hector jump? My inclination is that we would want the behavior to be consistent. 

`internal_Ftot = 3.7 * tgav / S + heatflux;  // inverting normal T=S/3.7*F eqn` where did this come from? is it consistent with doeclim?

**Comment by bpbond** [comment]
>where did this come from? is it consistent with doeclim?

It's just the mechanism that the temperature component used to use to allow it to exit 'gracefully' from a constraint. I don't think any other constraint does anything like this. Consistent with declaim—almost certainly not :)

**Comment by kdorheim** [comment]
If no other constraint has a graceful exit no need to add that complexity here which is great because um I have no idea how to back out the Ftot from the Tgav the doeclim way.

**Comment by bpbond** [comment]
Fixed in #558
